Plugins/GitUIPluginInterfaces/ObjectId.cs

```diff
@@ -0,0 +1,178 @@
+﻿using System;
+using System.Text;
+using JetBrains.Annotations;
+
+namespace GitUIPluginInterfaces
+{
+    /// <summary>
+    /// Models a SHA1 hash.
+    /// </summary>
+    /// <remarks>
+    /// Instances are immutable and are guaranteed to contain valid, 160-bit (20-byte) SHA1 hashes.
+    /// </remarks>
+    public sealed class ObjectId : IEquatable<ObjectId>
```

Comment:
Please separate state and behaviours
* `ObjectId` must be a POCO. `ToString`, `GetHashCode` etc are accepted since they are part of `Object` type definitions. `ToShortString` is ok too as it is complimentary to `ToString`
* All behavioural aspects, e.g. parsing must be in SOLID classes. Since all of them related to parsing it is probably natural to move them into `ObjectIdParser` type (or something along these lines).

`IsValid` bothers me as well, but I'm having troubles articulating where/how it should be moved.
I couldn't see - does it used anywhere besides tests?


And whilst you there, could you please group all fields, properties and methods together within respective groups. It makes the code much easier to review and reason about.

Thank you